topic_monitor/scripts/topic_monitor.py

```diff
@@ -0,0 +1,425 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import re
+from threading import Lock, Thread
+import time
+
+import rclpy
+from rclpy.qos import qos_profile_default, qos_profile_sensor_data
+
+from std_msgs.msg import Float32, Header
+
+QOS_DEPTH = 10
+
+
+class MonitoredTopic:
+    """Monitor for the statistics and status of a single topic."""
+
+    def __init__(self, topic_id, stale_time, lock=None):
+        self.topic_id = topic_id
+        self.stale_time = stale_time
+        self.status = 'Offline'
+        self.received_values = []
+        self.reception_rate_over_time = []
+        self.expected_value = None
+        self.time_of_last_data = None
+        self.initial_value = None
+        self.status_changed = False
+        self.expected_value_timer = None
+        self.lock = lock or Lock()
+
+    def increment_expected_value(self):
+        with self.lock:
+            if self.expected_value is not None:
+                self.expected_value += 1
+
+    def allowed_latency_timer_callback(self):
+        self.allowed_latency_timer.cancel()
+        self.expected_value_timer.reset()
+
+    def get_data_from_msg(self, msg):
+        data = msg.frame_id
+        return int(data[:data.find('_')])
+
+    def topic_data_callback(self, msg):
+        received_value = self.get_data_from_msg(msg)
+        print('%s: %s' % (self.topic_id, str(received_value)))
+        status = 'Alive'
+        with self.lock:
+            if self.expected_value is None:
+                # This is the first value from the topic
+                self.expected_value = received_value
+                self.initial_value = received_value - QOS_DEPTH
+                self.allowed_latency_timer.reset()
+            if received_value == -1:
+                # The topic was previously offline
+                status = 'Offline'
+                self.expected_value_timer.cancel()
+                self.expected_value = None
+            else:
+                self.received_values.append(received_value)
+            self.time_of_last_data = time.time()  # TODO: time stamp of msg
+            status_changed = status != self.status
+            self.status_changed |= status_changed  # don't clear the flag before check_status
+            self.status = status
+
+    def check_status(self, current_time=time.time()):
+        if self.status != 'Offline':
+            elapsed_time = current_time - self.time_of_last_data
+            if elapsed_time > self.stale_time * 1.1:
+                self.status_changed = self.status != 'Stale'
+                self.status = 'Stale'
+
+        status_changed = self.status_changed
+        self.status_changed = False
+        return status_changed
```

Comment:
The status could have been changed for other reasons: hopefully https://github.com/ros2/demos/pull/90/commits/7c0cb46099b4d7a9a8177c992818449bfaf4d6e0 clarifies this